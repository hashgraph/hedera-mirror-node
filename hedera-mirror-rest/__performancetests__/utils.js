#!/usr/bin/env node
/*-
 *
 * Hedera Mirror Node
 *
 * Copyright (C) 2019 Hedera Hashgraph, LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */

const fs = require('fs');
const yaml = require('js-yaml');

// Load config
process.env.CONFIG_NAME = 'perfTestConfig';
process.env.CONFIG_PATH = __dirname;
const config = require('../config');

const mustLoadYaml = fileName => {
  try {
    console.log(`Loading yaml file ${fileName}`);
    return yaml.safeLoad(fs.readFileSync(fileName, 'utf8'));
  } catch (err) {
    console.log(`Failed to load yaml file ${fileName}: ${err}`);
    process.exit(1);
  }
};

module.exports = {
  config: config,
  mustLoadYaml: mustLoadYaml
};
