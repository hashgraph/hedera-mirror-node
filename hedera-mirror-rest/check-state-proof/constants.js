/*-
 *
 * Hedera Mirror Node
 *  ​
 * Copyright (C) 2019 - 2021 Hedera Hashgraph, LLC
 *  ​
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */

'use strict';

const MAX_RECORD_LENGTH = 64 * 1024;

const MAX_TRANSACTION_LENGTH = 64 * 1024;

const SHA_384_LENGTH = 48;

const SHA_384_WITH_RSA = {
  type: 1,
  maxLength: 384,
};

const SIMPLE_SUM = 101;

module.exports = {
  MAX_RECORD_LENGTH,
  MAX_TRANSACTION_LENGTH,
  SHA_384_LENGTH,
  SHA_384_WITH_RSA,
  SIMPLE_SUM,
};
