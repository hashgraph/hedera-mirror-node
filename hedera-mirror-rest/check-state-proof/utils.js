/*-
 * ‌
 * Hedera Mirror Node
 * ​
 * Copyright (C) 2019 - 2020 Hedera Hashgraph, LLC
 * ​
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * ‍
 */
'uses strict';

// external libraries
const fs = require('fs');
const fetch = require('node-fetch');
const AbortController = require('abort-controller');

const base64StringToBuffer = (base64String) => {
  return Buffer.from(base64String, 'base64');
};

const replaceSpecialCharsWithUnderScores = (stringToFormat) => {
  return stringToFormat.replace(/[.@\-]/g, '_');
};

const makeStateProofDir = (transactionId, stateProofJson) => {
  const newDirPath = replaceSpecialCharsWithUnderScores(transactionId);
  fs.mkdirSync(newDirPath, {recursive: true});
  fs.writeFileSync(`${newDirPath}/apiResponse.json`, JSON.stringify(stateProofJson));
  console.log(`Supporting files and API response for the state proof will be stored in the directory ${newDirPath}`);
  return newDirPath;
};

const storeFile = (data, file, ext) => {
  const newFilePath = `${replaceSpecialCharsWithUnderScores(file)}.${ext}`;
  fs.writeFileSync(`${newFilePath}`, data, (err) => {
    if (err) throw err;
  });
};

const getAPIResponse = async (url) => {
  const controller = new AbortController();
  const timeout = setTimeout(
    () => {
      controller.abort();
    },
    60 * 1000 // in ms
  );

  console.log(`Requesting state proof files from ${url}...`);
  return await fetch(url, {signal: controller.signal})
    .then(async (response) => {
      if (!response.ok) {
        throw Error(response.statusText);
      }
      return response.json();
    })
    .catch((error) => {
      throw `Error fetching ${url}: ${error}`;
    })
    .finally(() => {
      clearTimeout(timeout);
    });
};

const readJSONFile = (filePath) => {
  const rawData = fs.readFileSync(filePath);
  return JSON.parse(rawData);
};

module.exports = {
  base64StringToBuffer,
  getAPIResponse,
  makeStateProofDir,
  readJSONFile,
  storeFile,
  replaceSpecialCharsWithUnderScores,
};
