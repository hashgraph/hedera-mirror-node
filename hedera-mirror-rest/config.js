/*-
 * ‌
 * Hedera Mirror Node
 *
 * Copyright (C) 2019 - 2020 Hedera Hashgraph, LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * ‍
 */

'use strict';

const extend = require('extend');
const fs = require('fs');
const yaml = require('js-yaml');
const path = require('path');
const {InvalidConfigError} = require('./errors/invalidConfigError');
const {cloudProviders, networks, defaultBucketNames} = require('./constants');

let configName = 'application';
if (process.env.CONFIG_NAME) {
  configName = process.env.CONFIG_NAME;
}

const config = {};
let loaded = false;

function load(configPath) {
  if (!configPath) {
    return;
  }

  let configFile = path.join(configPath, `${configName}.yml`);
  if (fs.existsSync(configFile)) {
    loadYaml(configFile);
  }

  configFile = path.join(configPath, configName + '.yaml');
  if (fs.existsSync(configFile)) {
    loadYaml(configFile);
  }
}

function loadYaml(configFile) {
  try {
    const doc = yaml.safeLoad(fs.readFileSync(configFile, 'utf8'));
    console.log(`Loaded configuration source: ${configFile}`);
    extend(true, config, doc);
  } catch (err) {
    console.log(`Skipping configuration ${configFile}: ${err}`);
  }
}

function loadEnvironment() {
  for (const [key, value] of Object.entries(process.env)) {
    setConfigValue(key, value);
  }
}

/*
 * Sets a config property from an environment variable by converting HEDERA_MIRROR_REST_FOO_BAR to an object path
 * notation hedera.mirror.rest.foo.bar using a case insensitive search. If more than one property matches with a
 * different case, it will choose the first.
 */
function setConfigValue(propertyPath, value) {
  let current = config;
  let properties = propertyPath.toLowerCase().split('_');

  // Ignore properties that don't start with HEDERA_MIRROR_REST
  if (properties.length < 4 || properties[0] !== 'hedera' || properties[1] !== 'mirror' || properties[2] !== 'rest') {
    return;
  }

  for (let i in properties) {
    let property = properties[i];
    let found = false;

    for (let [k, v] of Object.entries(current)) {
      if (property === k.toLowerCase()) {
        if (i < properties.length - 1) {
          current = v;
          found = true;
          break;
        } else {
          current[k] = convertType(value);
          console.log(`Override config with environment variable ${propertyPath}=${value}`);
          return;
        }
      }
    }

    if (!found) {
      return;
    }
  }
}

function convertType(value) {
  let parsedValue = value;

  if (value !== null && value !== '' && !isNaN(value)) {
    parsedValue = +value;
  } else if (value === 'true' || value === 'false') {
    parsedValue = value === 'true';
  }

  return parsedValue;
}

function getConfig() {
  return config.hedera && config.hedera.mirror ? config.hedera.mirror.rest : config;
}

function parseStateProofStreamsConfig() {
  const {stateproof} = getConfig();
  if (!stateproof || !stateproof.enabled) {
    return;
  }

  const {streams: streamsConfig} = stateproof;
  if (!Object.values(networks).includes(streamsConfig.network)) {
    throw new InvalidConfigError(`unknown network ${streamsConfig.network}`);
  }

  if (!streamsConfig.bucketName) {
    streamsConfig.bucketName = defaultBucketNames[streamsConfig.network];
  }

  if (!streamsConfig.bucketName) {
    // the default for network 'OTHER' is null, throw err if it's not configured
    throw new InvalidConfigError('stateproof.streams.bucketName must be set');
  }

  if (!Object.values(cloudProviders).includes(streamsConfig.cloudProvider)) {
    throw new InvalidConfigError(`unsupported object storage service provider ${streamsConfig.cloudProvider}`);
  }
}

if (!loaded) {
  load(path.join(__dirname, 'config'));
  load(__dirname);
  load(process.env.CONFIG_PATH);
  loadEnvironment();
  parseStateProofStreamsConfig();
  loaded = true;
}

module.exports = getConfig();
