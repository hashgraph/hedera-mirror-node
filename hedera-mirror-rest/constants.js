/*-
 * ‌
 * Hedera Mirror Node
 *​
 * Copyright (C) 2019 - 2021 Hedera Hashgraph, LLC
 *​
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */

'use strict';

// url query filer keys
const filterKeys = {
  ACCOUNT_ID: 'account.id',
  ACCOUNT_BALANCE: 'account.balance',
  ACCOUNT_PUBLICKEY: 'account.publickey',
  ENCODING: 'encoding',
  ENTITY_PUBLICKEY: 'publickey',
  LIMIT: 'limit',
  ORDER: 'order',
  RESULT: 'result',
  SEQUENCE_NUMBER: 'sequencenumber',
  TIMESTAMP: 'timestamp',
  TOKENID: 'tokenid',
  TOKEN_ID: 'token.id',
  CREDIT_TYPE: 'type',
  TRANSACTION_TYPE: 'transactiontype',
};

// sql table columns
const entityColumns = {
  ENTITY_NUM: 'entity_num',
  ENTITY_REALM: 'entity_realm',
  ENTITY_SHARD: 'entity_shard',
  PUBLIC_KEY: 'ed25519_public_key_hex',
};

const transactionColumns = {
  TYPE: 'type',
};

const requestIdLabel = 'requestId';
const responseDataLabel = 'mirrorRestData';

const orderFilterValues = {
  ASC: 'asc',
  DESC: 'desc',
};

// topic messages filter options
const characterEncoding = {
  BASE64: 'base64',
  UTF8: 'utf-8',
};

const transactionResultFilter = {
  SUCCESS: 'success',
  FAIL: 'fail',
};

const cryptoTransferType = {
  CREDIT: 'credit',
  DEBIT: 'debit',
};

const cloudProviders = {
  S3: 'S3',
  GCP: 'GCP',
};

const defaultCloudProviderEndpoints = {
  [cloudProviders.S3]: 'https://s3.amazonaws.com',
  [cloudProviders.GCP]: 'https://storage.googleapis.com',
};

const networks = {
  DEMO: 'DEMO',
  MAINNET: 'MAINNET',
  TESTNET: 'TESTNET',
  PREVIEWNET: 'PREVIEWNET',
  OTHER: 'OTHER',
};

const defaultBucketNames = {
  [networks.DEMO]: 'hedera-demo-streams',
  [networks.MAINNET]: 'hedera-mainnet-streams',
  [networks.TESTNET]: 'hedera-stable-testnet-streams-2020-08-27',
  [networks.PREVIEWNET]: 'hedera-preview-testnet-streams',
  [networks.OTHER]: null,
};

const recordStreamPrefix = 'recordstreams/record';

module.exports = {
  characterEncoding,
  cloudProviders,
  cryptoTransferType,
  defaultBucketNames,
  defaultCloudProviderEndpoints,
  entityColumns,
  filterKeys,
  networks,
  orderFilterValues,
  recordStreamPrefix,
  requestIdLabel,
  responseDataLabel,
  transactionColumns,
  transactionResultFilter,
};
