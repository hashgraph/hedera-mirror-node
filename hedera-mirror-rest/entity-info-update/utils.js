/*-
 * ‌
 * Hedera Mirror Node
 * ​
 * Copyright (C) 2019 - 2021 Hedera Hashgraph, LLC
 * ​
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * ‍
 */

'use strict';

// external libraries
const _ = require('lodash');
const fs = require('fs');
const log4js = require('log4js');
const math = require('mathjs');

// local
const config = require('./config');

const logger = log4js.getLogger();

const constructEntity = (index, headerRow, entityRow) => {
  const entityObj = {};
  const splitEntityRow = Array.from(entityRow.split(',')).filter((x) => x != null);

  for (let i = 0; i < headerRow.length; i++) {
    entityObj[headerRow[i].trim()] = splitEntityRow[i].trim();
  }

  return entityObj;
};

const readEntityCSVFileSync = () => {
  logger.info(`Parsing csv entity file ...`);
  const entities = [];

  const data = fs.readFileSync(config.filePath, 'utf-8');

  const fileContent = data.split('\n');
  const headers = Array.from(fileContent[0].split(',')).filter((x) => x != null);

  // ensure 1st column is entity num of expected format
  if (!_.eq(headers[0], 'entity')) {
    throw Error("CSV must have a header column with 1st column being 'entity'");
  }

  for (let i = 1; i < fileContent.length; i++) {
    if (!fileContent[i]) {
      // End of file
      break;
    }

    entities.push(constructEntity(i, headers, fileContent[i]));
  }

  logger.info(`${entities.length} entities were extracted from ${config.filePath}`);

  return entities;
};

/**
 * Converts seconds since epoch (seconds nnnnnnnnn format) to  nanoseconds
 * @param {String} Seconds since epoch (seconds.nnnnnnnnn format)
 * @return {String} ns Nanoseconds since epoch
 */
const secNsToNs = (secNs, ns) => {
  return math
    .add(math.multiply(math.bignumber(secNs.toString()), math.bignumber(1e9)), math.bignumber(ns.toString()))
    .toString();
};

module.exports = {
  readEntityCSVFileSync,
  secNsToNs,
};
