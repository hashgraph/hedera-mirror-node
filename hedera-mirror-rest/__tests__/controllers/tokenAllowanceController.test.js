/*-
 * ‌
 * Hedera Mirror Node
 * ​
 * Copyright (C) 2019 - 2023 Hedera Hashgraph, LLC
 * ​
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * ‍
 */

import {getResponseLimit} from '../../config';
import * as constants from '../../constants';
import * as utils from '../../utils';

const {default: defaultLimit} = getResponseLimit();
const {
  filterKeys: {SPENDER_ID, TOKEN_ID, ORDER, LIMIT},
} = constants;
const {
  opsMap: {eq, gt, gte, lt, lte, ne},
} = utils;

import Bound from '../../controllers/bound';
import {TokenAllowanceController} from '../../controllers';

const ownerAccountId = BigInt(1);

const spenderEqFilter = {key: SPENDER_ID, operator: eq, value: 5};
const spenderGtFilter = {key: SPENDER_ID, operator: gt, value: 6};
const spenderGteFilter = {key: SPENDER_ID, operator: gte, value: 7};
const spenderLtFilter = {key: SPENDER_ID, operator: lt, value: 15};
const spenderLteFilter = {key: SPENDER_ID, operator: lte, value: 16};

const tokenIdEqFilter = {key: TOKEN_ID, operator: eq, value: 100};
const tokenIdGtFilter = {key: TOKEN_ID, operator: gt, value: 101};
const tokenIdGteFilter = {key: TOKEN_ID, operator: gte, value: 102};
const tokenIdLtFilter = {key: TOKEN_ID, operator: lt, value: 150};
const tokenIdLteFilter = {key: TOKEN_ID, operator: lte, value: 151};

describe('extractTokenMultiUnionQuery', () => {
  const defaultExpected = {
    bounds: {
      primary: new Bound(SPENDER_ID, 'spender'),
      secondary: new Bound(TOKEN_ID, 'token_id'),
    },
    lower: [],
    inner: [],
    upper: [],
    ownerAccountId,
    order: constants.orderFilterValues.ASC,
    limit: defaultLimit,
  };

  const specs = [
    {
      name: 'empty',
      filters: [],
      expected: defaultExpected,
    },
    {
      name: 'order desc',
      filters: [{key: ORDER, operator: eq, value: 'desc'}],
      expected: {...defaultExpected, order: 'desc'},
    },
    {
      name: 'limit',
      filters: [{key: LIMIT, operator: eq, value: 60}],
      expected: {...defaultExpected, limit: 60},
    },
    {
      name: 'spender eq',
      filters: [spenderEqFilter],
      expected: {
        ...defaultExpected,
        bounds: {
          primary: Bound.create({equal: spenderEqFilter, filterKey: SPENDER_ID, viewModelKey: 'spender'}),
          secondary: new Bound(TOKEN_ID, 'token_id'),
        },
        lower: [spenderEqFilter],
      },
    },
    {
      name: 'spender gt',
      filters: [spenderGtFilter],
      expected: {
        ...defaultExpected,
        bounds: {
          primary: Bound.create({lower: spenderGtFilter, filterKey: SPENDER_ID, viewModelKey: 'spender'}),
          secondary: new Bound(TOKEN_ID, 'token_id'),
        },
        lower: [spenderGtFilter],
      },
    },
    {
      name: 'spender gte',
      filters: [spenderGteFilter],
      expected: {
        ...defaultExpected,
        bounds: {
          primary: Bound.create({lower: spenderGteFilter, filterKey: SPENDER_ID, viewModelKey: 'spender'}),
          secondary: new Bound(TOKEN_ID, 'token_id'),
        },
        lower: [spenderGteFilter],
      },
    },
    {
      name: 'spender lt',
      filters: [spenderLtFilter],
      expected: {
        ...defaultExpected,
        bounds: {
          primary: Bound.create({upper: spenderLtFilter, filterKey: SPENDER_ID, viewModelKey: 'spender'}),
          secondary: new Bound(TOKEN_ID, 'token_id'),
        },
        lower: [spenderLtFilter],
      },
    },
    {
      name: 'spender lte',
      filters: [spenderLteFilter],
      expected: {
        ...defaultExpected,
        bounds: {
          primary: Bound.create({upper: spenderLteFilter, filterKey: SPENDER_ID, viewModelKey: 'spender'}),
          secondary: new Bound(TOKEN_ID, 'token_id'),
        },
        lower: [spenderLteFilter],
      },
    },
    {
      name: 'spender gt and lt',
      filters: [spenderGtFilter, spenderLtFilter],
      expected: {
        ...defaultExpected,
        bounds: {
          primary: Bound.create({
            lower: spenderGtFilter,
            upper: spenderLtFilter,
            filterKey: SPENDER_ID,
            viewModelKey: 'spender',
          }),
          secondary: new Bound(TOKEN_ID, 'token_id'),
        },
        lower: [spenderGtFilter, spenderLtFilter],
      },
    },
    {
      name: 'spender eq and token eq',
      filters: [spenderEqFilter, tokenIdEqFilter],
      expected: {
        ...defaultExpected,
        bounds: {
          primary: Bound.create({equal: spenderEqFilter, filterKey: SPENDER_ID, viewModelKey: 'spender'}),
          secondary: Bound.create({equal: tokenIdEqFilter, filterKey: TOKEN_ID, viewModelKey: 'token_id'}),
        },
        lower: [spenderEqFilter, tokenIdEqFilter],
      },
    },
    {
      name: 'spender gt and token eq',
      filters: [spenderGtFilter, tokenIdEqFilter],
      expected: {
        ...defaultExpected,
        bounds: {
          primary: Bound.create({lower: spenderGtFilter, filterKey: SPENDER_ID, viewModelKey: 'spender'}),
          secondary: Bound.create({equal: tokenIdEqFilter, filterKey: TOKEN_ID, viewModelKey: 'token_id'}),
        },
        lower: [spenderGtFilter, tokenIdEqFilter],
      },
    },
    {
      name: 'spender lt and token eq',
      filters: [spenderLtFilter, tokenIdEqFilter],
      expected: {
        ...defaultExpected,
        bounds: {
          primary: Bound.create({upper: spenderLtFilter, filterKey: SPENDER_ID, viewModelKey: 'spender'}),
          secondary: Bound.create({equal: tokenIdEqFilter, filterKey: TOKEN_ID, viewModelKey: 'token_id'}),
        },
        lower: [spenderLtFilter, tokenIdEqFilter],
      },
    },
    {
      name: 'spender gt and lt, token eq',
      filters: [spenderGtFilter, spenderLtFilter, tokenIdEqFilter],
      expected: {
        ...defaultExpected,
        bounds: {
          primary: Bound.create({
            lower: spenderGtFilter,
            upper: spenderLtFilter,
            filterKey: SPENDER_ID,
            viewModelKey: 'spender',
          }),
          secondary: Bound.create({equal: tokenIdEqFilter, filterKey: TOKEN_ID, viewModelKey: 'token_id'}),
        },
        lower: [spenderGtFilter, spenderLtFilter, tokenIdEqFilter],
      },
    },
    {
      name: 'spender gte and lte, token eq',
      filters: [spenderGteFilter, spenderLteFilter, tokenIdEqFilter],
      expected: {
        ...defaultExpected,
        bounds: {
          primary: Bound.create({
            lower: spenderGteFilter,
            upper: spenderLteFilter,
            filterKey: SPENDER_ID,
            viewModelKey: 'spender',
          }),
          secondary: Bound.create({equal: tokenIdEqFilter, filterKey: TOKEN_ID, viewModelKey: 'token_id'}),
        },
        lower: [spenderGteFilter, spenderLteFilter, tokenIdEqFilter],
      },
    },
    {
      name: 'spender gte and token gte',
      filters: [spenderGteFilter, tokenIdGteFilter],
      expected: {
        ...defaultExpected,
        bounds: {
          primary: Bound.create({lower: spenderGteFilter, filterKey: SPENDER_ID, viewModelKey: 'spender'}),
          secondary: Bound.create({lower: tokenIdGteFilter, filterKey: TOKEN_ID, viewModelKey: 'token_id'}),
        },
        lower: [{...spenderGteFilter, operator: eq}, tokenIdGteFilter],
        inner: [{...spenderGteFilter, operator: gt}],
      },
    },
    {
      name: 'spender gte and token gt',
      filters: [spenderGteFilter, tokenIdGtFilter],
      expected: {
        ...defaultExpected,
        bounds: {
          primary: Bound.create({lower: spenderGteFilter, filterKey: SPENDER_ID, viewModelKey: 'spender'}),
          secondary: Bound.create({lower: tokenIdGtFilter, filterKey: TOKEN_ID, viewModelKey: 'token_id'}),
        },
        lower: [{...spenderGteFilter, operator: eq}, tokenIdGtFilter],
        inner: [{...spenderGteFilter, operator: gt}],
      },
    },
    {
      name: 'spender lte and token lte',
      filters: [spenderLteFilter, tokenIdLteFilter],
      expected: {
        ...defaultExpected,
        bounds: {
          primary: Bound.create({upper: spenderLteFilter, filterKey: SPENDER_ID, viewModelKey: 'spender'}),
          secondary: Bound.create({upper: tokenIdLteFilter, filterKey: TOKEN_ID, viewModelKey: 'token_id'}),
        },
        inner: [{...spenderLteFilter, operator: lt}],
        upper: [{...spenderLteFilter, operator: eq}, tokenIdLteFilter],
      },
    },
    {
      name: 'spender lte and token lt',
      filters: [spenderLteFilter, tokenIdLtFilter],
      expected: {
        ...defaultExpected,
        bounds: {
          primary: Bound.create({upper: spenderLteFilter, filterKey: SPENDER_ID, viewModelKey: 'spender'}),
          secondary: Bound.create({upper: tokenIdLtFilter, filterKey: TOKEN_ID, viewModelKey: 'token_id'}),
        },
        inner: [{...spenderLteFilter, operator: lt}],
        upper: [{...spenderLteFilter, operator: eq}, tokenIdLtFilter],
      },
    },
    {
      name: 'spender gte and lte, token gt',
      filters: [spenderGteFilter, spenderLteFilter, tokenIdGtFilter],
      expected: {
        ...defaultExpected,
        bounds: {
          primary: Bound.create({
            lower: spenderGteFilter,
            upper: spenderLteFilter,
            filterKey: SPENDER_ID,
            viewModelKey: 'spender',
          }),
          secondary: Bound.create({lower: tokenIdGtFilter, filterKey: TOKEN_ID, viewModelKey: 'token_id'}),
        },
        lower: [{...spenderGteFilter, operator: eq}, tokenIdGtFilter],
        inner: [{...spenderGteFilter, operator: gt}, spenderLteFilter],
      },
    },
    {
      name: 'spender gte and lte, token lt',
      filters: [spenderGteFilter, spenderLteFilter, tokenIdLtFilter],
      expected: {
        ...defaultExpected,
        bounds: {
          primary: Bound.create({
            lower: spenderGteFilter,
            upper: spenderLteFilter,
            filterKey: SPENDER_ID,
            viewModelKey: 'spender',
          }),
          secondary: Bound.create({upper: tokenIdLtFilter, filterKey: TOKEN_ID, viewModelKey: 'token_id'}),
        },
        inner: [
          {...spenderGteFilter, operator: gte},
          {...spenderLteFilter, operator: lt},
        ],
        upper: [{...spenderLteFilter, operator: eq}, tokenIdLtFilter],
      },
    },
    {
      name: 'spender gte and lte, token gt and lt',
      filters: [spenderGteFilter, spenderLteFilter, tokenIdGtFilter, tokenIdLtFilter],
      expected: {
        ...defaultExpected,
        bounds: {
          primary: Bound.create({
            lower: spenderGteFilter,
            upper: spenderLteFilter,
            filterKey: SPENDER_ID,
            viewModelKey: 'spender',
          }),
          secondary: Bound.create({
            lower: tokenIdGtFilter,
            upper: tokenIdLtFilter,
            filterKey: TOKEN_ID,
            viewModelKey: 'token_id',
          }),
        },
        lower: [{...spenderGteFilter, operator: eq}, tokenIdGtFilter],
        inner: [
          {...spenderGteFilter, operator: gt},
          {...spenderLteFilter, operator: lt},
        ],
        upper: [{...spenderLteFilter, operator: eq}, tokenIdLtFilter],
      },
    },
    {
      name: 'spender gte and lte, token gte and lte',
      filters: [spenderGteFilter, spenderLteFilter, tokenIdGteFilter, tokenIdLteFilter],
      expected: {
        ...defaultExpected,
        bounds: {
          primary: Bound.create({
            lower: spenderGteFilter,
            upper: spenderLteFilter,
            filterKey: SPENDER_ID,
            viewModelKey: 'spender',
          }),
          secondary: Bound.create({
            lower: tokenIdGteFilter,
            upper: tokenIdLteFilter,
            filterKey: TOKEN_ID,
            viewModelKey: 'token_id',
          }),
        },
        lower: [{...spenderGteFilter, operator: eq}, tokenIdGteFilter],
        inner: [
          {...spenderGteFilter, operator: gt},
          {...spenderLteFilter, operator: lt},
        ],
        upper: [{...spenderLteFilter, operator: eq}, tokenIdLteFilter],
      },
    },
  ];

  specs.forEach((spec) => {
    test(spec.name, () => {
      expect(TokenAllowanceController.extractTokenMultiUnionQuery(spec.filters, ownerAccountId)).toEqual(spec.expected);
    });
  });
});

describe('extractTokenMultiUnionQuery throw', () => {
  const specs = [
    {
      name: 'spender.id ne',
      filters: [{key: SPENDER_ID, operator: ne, value: 1}],
    },
    {
      name: 'spender.id range and equal',
      filters: [spenderEqFilter, spenderGtFilter, spenderLtFilter],
    },
    {
      name: 'spender.id multiple equal',
      filters: [spenderEqFilter, spenderEqFilter],
    },
    {
      name: 'spender.id multiple lower bounds',
      filters: [spenderGtFilter, spenderGteFilter],
    },
    {
      name: 'spender.id multiple upper bounds',
      filters: [spenderLtFilter, spenderLteFilter],
    },
    {
      name: 'token.id ne',
      filters: [{key: TOKEN_ID, operator: ne, value: 1}],
    },
    {
      name: 'token.id range and equal',
      filters: [tokenIdEqFilter, tokenIdGtFilter, tokenIdLtFilter],
    },
    {
      name: 'token.id multiple equal',
      filters: [tokenIdEqFilter, tokenIdEqFilter],
    },
    {
      name: 'token.id multiple lower bounds',
      filters: [tokenIdGtFilter, tokenIdGteFilter],
    },
    {
      name: 'token.id multiple upper bounds',
      filters: [tokenIdLtFilter, tokenIdLteFilter],
    },
    {
      name: 'token.id eq without spender.id query',
      filters: [tokenIdEqFilter],
    },
    {
      name: 'token.id gt without spender.id query',
      filters: [tokenIdGtFilter],
    },
    {
      name: 'token.id gt with spender.id gt',
      filters: [tokenIdGtFilter, spenderGtFilter],
    },
    {
      name: 'token.id lt without spender.id query',
      filters: [tokenIdLtFilter],
    },
    {
      name: 'token.id lt with spender.id lt',
      filters: [tokenIdLtFilter, spenderLtFilter],
    },
  ];

  specs.forEach((spec) => {
    test(spec.name, () => {
      expect(() =>
        TokenAllowanceController.extractTokenMultiUnionQuery(spec.filters, ownerAccountId)
      ).toThrowErrorMatchingSnapshot();
    });
  });
});
