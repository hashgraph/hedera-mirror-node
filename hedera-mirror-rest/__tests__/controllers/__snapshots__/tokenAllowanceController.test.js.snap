// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`extractTokenMultiUnionQuery throw spender.id multiple equal 1`] = `"Only one equal (eq) operator is allowed for spender.id"`;

exports[`extractTokenMultiUnionQuery throw spender.id multiple lower bounds 1`] = `"Only one gt/gte operator is allowed for spender.id"`;

exports[`extractTokenMultiUnionQuery throw spender.id multiple upper bounds 1`] = `"Only one lt/lte operator is allowed for spender.id"`;

exports[`extractTokenMultiUnionQuery throw spender.id ne 1`] = `"Not equal (ne) operator is not supported for spender.id"`;

exports[`extractTokenMultiUnionQuery throw spender.id range and equal 1`] = `"Can't support both range and equal for spender.id"`;

exports[`extractTokenMultiUnionQuery throw token.id eq without spender.id query 1`] = `"token.id without a spender.id parameter filter"`;

exports[`extractTokenMultiUnionQuery throw token.id gt with spender.id gt 1`] = `"spender.id must have gte or eq operator"`;

exports[`extractTokenMultiUnionQuery throw token.id gt without spender.id query 1`] = `"token.id without a spender.id parameter filter"`;

exports[`extractTokenMultiUnionQuery throw token.id lt with spender.id lt 1`] = `"spender.id must have lte or eq operator"`;

exports[`extractTokenMultiUnionQuery throw token.id lt without spender.id query 1`] = `"token.id without a spender.id parameter filter"`;

exports[`extractTokenMultiUnionQuery throw token.id multiple equal 1`] = `"Only one equal (eq) operator is allowed for token.id"`;

exports[`extractTokenMultiUnionQuery throw token.id multiple lower bounds 1`] = `"Only one gt/gte operator is allowed for token.id"`;

exports[`extractTokenMultiUnionQuery throw token.id multiple upper bounds 1`] = `"Only one lt/lte operator is allowed for token.id"`;

exports[`extractTokenMultiUnionQuery throw token.id ne 1`] = `"Not equal (ne) operator is not supported for token.id"`;

exports[`extractTokenMultiUnionQuery throw token.id range and equal 1`] = `"Can't support both range and equal for token.id"`;
